mne/preprocessing/maxwell.py

```diff
@@ -0,0 +1,507 @@
+        position info must be in the same coordinate frame as 'origin'
+    int_order : int
+        Order of the internal multipolar moment space
+    ext_order : int
+        Order of the external multipolar moment space
+
+    Returns
+    -------
+    tuple, len (2)
+        Internal and external basis sets ndarrays with shape
+        (n_coils, n_mult_moments)
+    """
+    r_int_pts, ncoils, wcoils, int_pts = _concatenate_coils(coils)
+    n_sens = len(int_pts)
+    n_bases = get_num_moments(int_order, ext_order)
+    int_lens = np.insert(np.cumsum(int_pts), obj=0, values=0)
+
+    S_in = np.empty((n_sens, (int_order + 1) ** 2 - 1))
+    S_out = np.empty((n_sens, (ext_order + 1) ** 2 - 1))
+    S_in.fill(np.nan)
+    S_out.fill(np.nan)
+
+    # Set all magnetometers (with 'coil_type' == 1.0) to be scaled by 100
+    coil_scale = np.ones((len(coils)))
+    coil_scale[np.array([coil['coil_class'] == 1.0 for coil in coils])] = 100.
+
+    assert n_bases <= n_sens, ('Number of requested bases (%s) exceeds number '
+                               'of sensors (%s)' % (str(n_bases), str(n_sens)))
+
+    # Compute position vector between origin and coil integration pts
+    cvec_cart = r_int_pts - origin[np.newaxis, :]
+    # Convert points to spherical coordinates
+    cvec_sph = _cart_to_sph(cvec_cart)
+
+    # Compute internal basis vectors (exclude degree 0; RHS Eq. 5)
+    for deg in range(1, int_order + 1):
+        for order in range(-deg, deg + 1):
+
+            # Compute gradient for integration point position vectors
+            grads = -1 * _grad_in_components(deg, order, cvec_sph[:, 0],
+                                             cvec_sph[:, 1], cvec_sph[:, 2])
+
+            # Gradients dotted with integration point normals and weighted
+            a1_all = wcoils * np.einsum('ij,ij->i', grads, ncoils)
+
+            # For order and degree, sum across integration pts for each sensor
+            for pt_i in range(0, len(int_lens) - 1):
+                int_pts_sum = np.sum(a1_all[int_lens[pt_i]:int_lens[pt_i + 1]])
+                S_in[pt_i, deg ** 2 + deg + order - 1] = int_pts_sum
+
+    # Compute external basis vectors (exclude degree 0; RHS Eq. 5)
+    for deg in range(1, ext_order + 1):
+        for order in range(-deg, deg + 1):
+
+            # Compute gradient for integration point position vectors
+            grads = -1 * _grad_out_components(deg, order, cvec_sph[:, 0],
+                                              cvec_sph[:, 1], cvec_sph[:, 2])
+
+            # Gradients dotted with integration point normals and weighted
+            b1_all = wcoils * np.einsum('ij,ij->i', grads, ncoils)
+
+            # For order and degree, sum across integration pts for each sensor
+            for pt_i in range(0, len(int_lens) - 1):
+                int_pts_sum = np.sum(b1_all[int_lens[pt_i]:int_lens[pt_i + 1]])
+                S_out[pt_i, deg ** 2 + deg + order - 1] = int_pts_sum
```

Comment:
The only difference between this and the above is the `_grad_*_` function and the destination array, right? So you can do:
```
for out, fun in zip([S_in, S_out], [_grad_in_components, _grad_out_components]):
    ...
```
and DRY your code, right?